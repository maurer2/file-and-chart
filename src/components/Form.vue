<template>
  <form class="form" action="/" method="post">
    <label class="label" for="field">Upload a CSV-file</label>
    <input
      class="field"
      type="file"
      name="field"
      id="field"
      @change="handleChangeEvent"
    >
    <div class="button-group">
      <button class="button" type="button" @click="openCatsCSV">
        Cats
      </button>
      <button class="button" type="button" @click="openDogsCSV">
        Dogs
      </button>
    </div>
  </form>
</template>

<script>
import catsCSV from '../../csv/cats.csv';
import dogsCSV from '../../csv/dogs.csv';

export default {
  name: 'Form',
  props: {},
  methods: {
    handleChangeEvent(event) {
      const [file] = event.target.files;

      if (file.type !== 'text/csv') {
        return;
      }

      this.getFileContent(file);
    },
    getFileContent(file) {
      const fileReader = new FileReader();

      fileReader.onload = () => {
        const results = fileReader.result;

        this.$emit('fileUpload', results);
      };

      fileReader.readAsText(file);
    },
    openCatsCSV() {
      this.$emit('fileUpload', catsCSV);
    },
    openDogsCSV() {
      this.$emit('fileUpload', dogsCSV);
    },
  },
};
</script>

<style scoped lang="scss">
.form {
  position: relative;
  padding: 0;
}

.label {
  display: inline-block;
  padding: 0.5rem 1rem;
  background: whitesmoke;
  border: 1px solid #d8d8d8;
}

// hide field visually
.field {
  position: absolute;
  left: -9999px;
  opacity: 0.01;
}

.button-group {
  display: flex;
  margin-top: 1rem;
  justify-content: center;
}

.button {
  margin: 0 0.5rem;
  padding: 0.25rem 0.5rem;
  background: whitesmoke;
}

</style>
