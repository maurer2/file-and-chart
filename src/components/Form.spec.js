import { shallowMount } from '@vue/test-utils';
import Form from '@/components/Form.vue';

describe('Form', () => {
  let wrapper;

  const mockedChangeEvent = jest.fn();
  const mockedGetFileContent = jest.fn();

  beforeEach(() => {
    wrapper = shallowMount(Form, {
      // methods: { handleChangeEvent: mockedChangeEvent },
      attachToDocument: true,
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('Form snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Form contains all elements and child elements', () => {
    expect(wrapper.find('.form').exists()).toBe(true);
    expect(wrapper.find('.label').exists()).toBe(true);
    expect(wrapper.find('.field').exists()).toBe(true);
    expect(wrapper.find('.button-group .button:nth-child(1)').exists()).toBe(true);
    expect(wrapper.find('.button-group .button:nth-child(2)').exists()).toBe(true);
  });

  test('Form field is of type file', () => {
    expect(wrapper.find('.field').attributes('type')).toBe('file');
    expect(wrapper.find('.field').attributes('type')).not.toBe('text');
  });

  test('Form field id is the same as label for attribute', () => {
    const fieldId = wrapper.find('.field').attributes('id');

    expect(wrapper.find('.label').attributes('for')).toBe(fieldId);
  });

  test('Form field change event triggers handleChangeEvent', () => {
    wrapper.setMethods({ handleChangeEvent: mockedChangeEvent });
    wrapper.find('.field').trigger('change');

    expect(mockedChangeEvent).toBeCalled();
  });

  test('Form field handleChangeEvent calls getFileContent when correct file is set', () => {
    wrapper.setMethods({ getFileContent: mockedGetFileContent });

    wrapper.vm.handleChangeEvent(new Event('change'));
    expect(mockedGetFileContent).not.toBeCalled();
  });

  test.skip('Form getFileContent emits event when correct file is uploaded', (done) => {
    const testBlob = new Blob(['cat'], { type: 'text/plain' });

    wrapper.vm.getFileContent(testBlob);

    wrapper.vm.$nextTick(() => {
      expect(wrapper.emitted('fileUpload')).toBeTruthy();
      done();
    });
  });

  test('Form openCatsCSV emits event', () => {
    wrapper.vm.openCatsCSV();

    expect(wrapper.emitted('fileUpload')).toBeTruthy();
  });

  test('Form openDogsCSV emits event', () => {
    wrapper.vm.openCatsCSV();

    expect(wrapper.emitted('fileUpload')).toBeTruthy();
  });
});
