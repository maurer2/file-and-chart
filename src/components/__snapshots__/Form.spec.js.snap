// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Form snapshot 1`] = `
<form action="/" method="post" class="form"><label for="field" class="label">Upload a CSV-file</label> <input type="file" name="field" id="field" class="field">
  <div class="button-group"><button type="button" class="button">
      Cats
    </button> <button type="button" class="button">
      Dogs
    </button></div>
</form>
`;
